File: runtime/java/units/memory/copy.java
// memory.copy must be overlap-safe; System.arraycopy is, like memmove.
void copy(long dst, long src, long length) {
    int da = at(dst, length);
    int sa = at(src, length);
    System.arraycopy(d, sa, d, da, (int) length);
}

File: runtime/perl/units/memory/copy.pl
# 4-arg substr copies the extracted source before replacing, so overlapping ranges are safe.
sub copy {
    my ($self, $dst, $src, $len) = @_;
    $self->check($dst, $len);
    $self->check($src, $len);
    return if $len == 0;
    substr($self->{data}, $dst, $len, substr($self->{data}, $src, $len));
}
